### PR TITLE
rule: no-private-properties

### LINKED ISSUE #30: Rule to check that private property of custom element is not being passed to element
Just started using LitElement / LitHtml after mainly using AngularJS and Angular 2+ in my previous assignment.

From what I understand, besides properties used to supply data from the caller template you can also define a property that is only used internally in your component.
To indicate that property is internal to the component, I prefixed it's name with a `_` and also added the `private` access modifier to its `@property` decorated field.

Wouldn't it be cool if ESLint lit plugin could lint for property assignment in lit-html templates
that start with `._`?
Or am I doing it all wrong and should I just stop (mis-)using `@property` for private properties?

### PR DESCRIPTION
**Pull request by stramel**
Fixes #30 

Currently based on the update-scripts PR so my life would be easier while developing. Will rebase once that is merged.

### CODE REVIEW

File: src/rules/no-private-properties.ts [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+        },
+        additionalProperties: false
+      }
+    ]
+  },
+
+  create(context): Rule.RuleListener {
+    // variables should be defined here
+    const options = {
+      prefix: '_',
```
Comment (43081j) [inline src/rules/no-private-properties.ts]: i'd rather this _always_ be a regex and just `new RegExp(prefix)` it somewhere here and re-use that (so the default is `^_`)

File: src/rules/no-private-properties.ts [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+              // eslint-disable-next-line guard-for-in
+              for (const attr in element.attribs) {
+                const loc = analyzer.getLocationForAttribute(element, attr);
+
+                if (!loc) {
+                  continue;
+                }
+
+                const hasBinding = bindings.includes(attr.slice(0, 1));
+                const normalizedAttr = hasBinding ? attr.slice(1) : attr;
```
Comment (43081j) [inline src/rules/no-private-properties.ts]: we should only support `.` as the rest are attributes or events and those can be named whatever you like (we don't have a convention of "private events" or attributes).

File: src/rules/no-private-properties.ts (hunk 1 above)
Comment (stramel) [reply src/rules/no-private-properties.ts]: Yeah, going to change to follow the `"@typescript-eslint/member-naming"` with options:

```
      {
        "private": "^__",
        "protected": "^_"
      }
```

File: docs/rules/no-private-properties.md [hunk 3]
```diff
@@ -0,0 +1,40 @@
+# Detects usages of non-public property bindings (no-private-properties)
+
+It can be helpful to enforce naming conventions for `private` (and sometimes `protected`) members of an object. For example, prefixing private properties with a `__` allows them to be easily discerned when being inspected by tools that do not have knowledge of TypeScript (such as most debuggers).
+
+## Rule Details
+
+This rule allows you to enforce conventions for property names by their visibility. By default, it enforces nothing.
```
Comment (43081j) [inline docs/rules/no-private-properties.md]: We should reword this to make it clear we're not testing property names in general but rather disallowing non-public ones from being set.

Maybe:

> This rule allows you to enforce that all properties being set on elements are public.

File: docs/rules/no-private-properties.md [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+## Options
+
+You can specify a regular expression to test the names of properties for each visibility level: `protected` and/or `private`.
```
Comment (43081j) [inline docs/rules/no-private-properties.md]: We should reword this one too, possibly:

> You can specify a regular expression to be tested against when detecting the visibility of a property. `private` and `protected` are supported.

File: docs/rules/no-private-properties.md [hunk 5]
```diff
@@ -0,0 +1,40 @@
+# Detects usages of non-public property bindings (no-private-properties)
```
Comment (43081j) [inline docs/rules/no-private-properties.md]: This should match up with whats in the header of each file

File: src/rules/no-private-properties.ts [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+        },
+        additionalProperties: false,
+        minProperties: 1
+      }
+    ]
+  },
+
+  create(context): Rule.RuleListener {
+    // variables should be defined here
+    const config: {private?: string; protected?: string} =
```
Comment (43081j) [inline src/rules/no-private-properties.ts]: a `Partial<{private: string; protected: string}>` is preferred so we don't overmatch

File: src/rules/no-private-properties.ts [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+      context.options[0] || {};
+    const conventions = Object.entries(config).reduce(
```
Comment (43081j) [inline src/rules/no-private-properties.ts]: you can probably `reduce<Record<string, RegExp>>` and remove the ` as `

File: docs/rules/no-private-properties.md [hunk 8]
```diff
@@ -0,0 +1,40 @@
+# Disallows usages of "non-public" property bindings (no-private-properties)
+
+It can be helpful to enforce naming conventions for `private` (and sometimes `protected`) members of an object. For example, prefixing private properties with a `__` allows them to be easily discerned when being inspected by tools that do not have knowledge of TypeScript (such as most debuggers).
```
Comment (43081j) [inline docs/rules/no-private-properties.md]: this should probably be across multiple lines

Comment (43081j) [reply docs/rules/no-private-properties.md]: to be honest i would probably reword to put emphasis on what this rule does, too, rather than about why someone might want visibility naming.

> When following a naming convention to signify the visibility of properties, any which are non-public should
> not be accessed by a consumer of the element. Such properties should be made public and documented
> as such. For example, when following the convention of private properties being prefixed
> by `__` (two underscores) and protected properties being prefixed by `_` (one underscore), it
> should generally be required that only properties without a `__` or `_` prefix are accessed
> from outside the element.

Comment (stramel) [reply docs/rules/no-private-properties.md]: Will do this when I get home
